prow/cmd/build/controller_test.go

```diff
@@ -214,6 +235,377 @@ func TestEnqueueKey(t *testing.T) {
 	}
 }
 
+func TestTerminateDupProwJobs(t *testing.T) {
```

Comment:
What happens if we fail to delete the build on the first pass, but succeed in marking the ProwJob as aborted? Do we retry to delete the build? Do we leak them?

Reply:
Yes, this is a good point. In the current implementation, the job is marked as aborted whether or not the resources are successfully cleaned up. The error is only logged and these resources will be eventually leaked.  

I would suggest that we skip marking the job as aborted if the resources cannot be cleaned up. This means the delete can be retried in the next reconciliation loop. 

@fejta What is your opinion on this?

Reply:
Ideally we should be level-based, rather than edge-based, and the system should recover from this situation on subsequent passes.

The details get kind of complicated, but fundamentally, the pattern here ought to be independent systems:

* System A looks for old prow jobs and aborts the job
  - Assuming we're configured to do this
* System B looks for aborted jobs that have a running pod/build and terminates the pod/build
  - Depending on whether the pod/build is configured to continue or terminate on job abort
* System C looks for prow jobs waiting to start a pod/build and starts them
  - minding concurrency restrictions for the job
  - counting zombie pods/builds associated with an aborted job towards the concurrency limit

AKA, we will continually attempt to 
* abort jobs that are not the latest for that commit (if configured to do this)
* delete pods/builds for aborted jobs (if the job allows this)
* start new pods/builds (as old pods/builds finish)

However I will be surprised if plank gets any of this behavior exactly right today. I would set the goal of this PR to:
* Not modify plank's behavior at all
* Make build and plank have matching behavior

Reply:
@fejta Thanks! These details are great!

> However I will be surprised if plank gets any of this behavior exactly right today. I would set the goal of this PR to:
> - Not modify plank's behavior at all
> - Make build and plank have matching behavior

My goal was to preserve the plank behavior as it is today to avoid introducing any regression.

I added a TODO where the `cleanup` is called and linked to this discussion. I would tackle this in a follow up PR.